packages/ag-charts-community/src/chart/legend/legendManager.ts

```diff
@@ -0,0 +1,151 @@
+import type { AgInitialStateLegendOptions } from 'ag-charts-types';
+
+import type { MementoOriginator } from '../../api/state/memento';
+import { isArray } from '../../util/type-guards';
+import { BaseManager } from '../baseManager';
+import type { ChartService } from '../chartService';
+import type { CategoryLegendDatum } from './legendDatum';
+
+export interface LegendChangeEvent {
+    type: 'legend-change';
+    legendData?: CategoryLegendDatum[];
+}
+
+type LegendDataMap = Map<string, CategoryLegendDatum[]>;
+export type LegendDataMemento = AgInitialStateLegendOptions[];
+
+export class LegendManager
+    extends BaseManager<LegendChangeEvent['type'], LegendChangeEvent>
+    implements MementoOriginator<LegendDataMemento>
+{
+    public constructor(readonly chartService: ChartService) {
+        super();
+    }
+    public mementoOriginatorKey = 'legend' as const;
+
+    private readonly legendDataMap: LegendDataMap = new Map();
+
+    public createMemento() {
+        return this.getData().map(({ enabled, ...datum }) => ({ ...datum, visible: enabled }));
+    }
+
+    public guardMemento(blob: unknown): blob is LegendDataMemento | undefined {
+        return blob == null || isArray(blob);
+    }
+
+    public restoreMemento(_version: string, _mementoVersion: string, memento: LegendDataMemento | undefined) {
+        const { legendDataMap } = this;
+        memento?.forEach((datum) => {
+            const { seriesId, data } = this.getRestoredDatum(datum) ?? {};
+
+            if (!(seriesId && data)) {
```

Comment:
Nested statements like this are a bit confusing when skimming through. I'd suggest avoiding them in favour of `if (!seriesId || !data) {`